content/blog/2025-12-16-loop-parallelization-proposal.md.md

```diff
@@ -0,0 +1,552 @@
+
+With memory bandwidth saturated, adding more threads or reducing overhead doesn't help. In fact, fusion can hurt by:
+- Increasing working set size (processing more arrays simultaneously)
+- Reducing cache hit rates
+- Creating more cache conflicts
+
+## 2. Parallel Region Overhead is Negligible
+
+Modern OpenMP implementations (LLVM's libomp) are highly optimized. The overhead of creating parallel regions is on the order of microseconds. For our benchmarks:
+
+```
+Typical loop execution time: ~10-100ms
+Parallel region overhead: ~1-10μs
+Overhead percentage: < 0.1%
+```
+
+Eliminating this overhead through shared regions provides minimal benefit.
+
+## 3. Fusion Introduces Complexity
+
+Loop fusion adds several sources of overhead:
+
+**Compiler overhead:**
+- More complex dependence analysis
+- Additional IR transformations
+- Larger basic blocks
+
+**Runtime overhead:**
+- Fused loops have larger loop bodies
+- Reduces instruction cache efficiency
+- Can prevent other optimizations (e.g., loop unrolling)
+
+**Example:** Fusing two 5-instruction loops creates a 10-instruction loop body, potentially:
+- Exceeding instruction cache line size
+- Preventing SIMD vectorization
+- Increasing register pressure
+
+## 4. Small Problem Sizes
+
+PolyBench uses `SMALL_DATASET` configuration:
+- Matrix sizes: typically 100x100 to 400x400
+- Total operations: 10K - 1M
+- Execution time per loop: < 1ms in many cases
+
+**Impact:** For such small problems:
+- Serial execution is very fast (microseconds)
+- Parallelization overhead dominates
+- No opportunity for fusion benefits to manifest
+
+## 5. Limited Optimization Space in Real-World Benchmarks
+
+The PolyBench benchmarks revealed a fundamental limitation: **only 20% of loops (6 out of 30 benchmarks) were parallelizable** using our conservative analysis.
+
+**Why most loops couldn't be parallelized:**
+
+Many PolyBench kernels have inherent sequential dependencies that prevent parallelization:
+
+```c
+// Example: Jacobi stencil (sequential dependency)
+for (int t = 0; t < tsteps; t++)           // Outer time-step loop
+    for (int i = 1; i < n-1; i++)
+        for (int j = 1; j < n-1; j++)
+            B[i][j] = (A[i][j] + A[i-1][j] + A[i+1][j] +
+                       A[i][j-1] + A[i][j+1]) * 0.2;
+```
+
+The outer time-step loop creates dependencies: iteration `t` depends on results from iteration `t-1`. While the inner spatial loops could be parallelized, our pass currently doesn't handle nested loop parallelization.
+
+**Limited fusion opportunities:**
+
+Even among the 6 parallelizable benchmarks, fusion opportunities were rare because:
+1. **Most benchmarks have single dominant loops** rather than sequences of similar loops
+2. **Different trip counts**: Consecutive loops often iterate over different dimensions (e.g., one over rows, next over columns)
+3. **Data dependencies**: Even when structurally similar, loops often have dependencies (e.g., first loop computes values used by second loop)
```

Comment:
This section is about limited parallelization opportunities, not about fusion opportunities. This just seems to confuse the issue.